File: winter2009/comp2150/a5/q2/lib/customer_departure_event.rb
# file: lib/customer_departure_event.rb

# Models the departure of a customer from the store simulation.
class CustomerDepartureEvent < DepartureEvent

  def initialize(time, customer)
    super(time, customer)
    @customer = customer
  end

  # Event action. Remove customer from checkout, and tell checkout to advance next
  # customer in line. Record departure time for stats.
  def run
    SoopaStore.checkout.occupied = false
    SoopaStore.checkout.start_next_checkout
    @customer.departure_time = SoopaStore.current_time
    puts "Time #{SoopaStore.current_time}  : Customer  #{@customer.customer_id} "\
    "departs (spent #{@customer.time_spent_waiting} units waiting)"
  end
  
end

File: winter2009/comp2150/a2/CustomerDepartureEvent.java
// CLASS: CustomerDepartureEvent
//
// REMARKS: Models the departure of a customer from an event-based simulation
//
// Input: None
//
// Output: None
//
//-----------------------------------------

// CustomerDepartureEvent:
//   if |Q| > 0:
//     T Q.front -> StartCheckoutEvent @ t+1

public class CustomerDepartureEvent extends DepartureEvent {

  //------------------------------------------------------
  // CustomerDepartureEvent
  //
  // PURPOSE: Constructs a new CustomerDepartureEvent
  // PARAMETERS:
  //     time: time at which event occurs
  //     customer: customer departing
  //     checkout: checkout customer is departing from
  // EXTERNAL REFERENCES:  requires Customer, Checkout
  //------------------------------------------------------
   public CustomerDepartureEvent(int time, Customer customer, Checkout checkout) {
    super(time,customer,checkout);
  }
  
  //------------------------------------------------------
  // action
  //
  // PURPOSE: called when this event "happens". If the line behind them is
  //     not empty, the customer will trigger a StartCheckoutEvent on the front.
  // EXTERNAL REFERENCES: requires StartCheckoutEvent, Customer
  //------------------------------------------------------
  public void action() {
    Customer customer = (Customer) this.getActor();

    checkout.setCheckingOut((Customer) null);
    
    if (checkout.getLength() > 0) {
      Event event = new StartCheckoutEvent(time+1, checkout.firstInLine(), this.checkout);
      this.store.getEventList().addEvent(event);
    }

    checkout.setCheckingOut(checkout.firstInLine());
    this.checkout.moveLine();

    System.out.printf("Time %d: Customer %d departs (spent %d units waiting)\n",
                      time,
                      customer.getId(),
                      0);
  }
  
}